src/Markup/Avalonia.Markup.Xaml/SourceInfo/SourceInfo.cs

```diff
@@ -0,0 +1,215 @@
+﻿using System;
+using Avalonia.Controls;
+
+namespace Avalonia.Markup.Xaml.SourceInfo
+{
+    /// <summary>
+    /// Represents source location information for an element within a XAML or code file.
+    /// </summary>
+    /// <remarks>
+    /// This struct is typically used to store the line, column, and file path of a control or object
+    /// to enable navigation or diagnostics (for example, jumping to a definition in Visual Studio).
+    /// </remarks>
+    public readonly struct SourceInfo : IEquatable<SourceInfo>
```

Comment:
This is more me thinking aloud than anything else, but I'm not sure it's a good idea for this to be a `struct`? It limits what we can do with this type, and it's going to be boxed anyway if stored in an attached property.